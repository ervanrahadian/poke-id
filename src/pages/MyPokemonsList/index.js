import React, { useEffect, useState } from "react";
import { Col, Row } from "react-bootstrap";
import { Microtransaction, MyListCard, Toast } from "../../components";

/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";

const MyPokemonsList = () => {
  const [pokecoin, setPokecoin] = useState(null);
  const [myListData, setMyListData] = useState(null);

  useEffect(() => {
    if (localStorage.hasOwnProperty("Pokecoin")) {
      setPokecoin(Number(localStorage.getItem("Pokecoin")));
    } else {
      const defaultPokecoin = Number(500);

      localStorage.setItem("Pokecoin", defaultPokecoin);
      setPokecoin(defaultPokecoin);

      return Toast({
        icon: "success",
        title: "500 Pokecoin added!",
      });
    }

    if (localStorage.hasOwnProperty("myPoke")) {
      const check = localStorage.getItem("myPoke");

      if (check) {
        setMyListData(JSON.parse(localStorage.getItem("myPoke")));
      }
    }
  }, []);

  const style = {
    mainTitle: css({
      fontSize: "32px",
      fontWeight: "800",
    }),
    text: css({
      fontSize: "24px",
      fontWeight: "500",
    }),
  };

  return (
    <>
      <Microtransaction pokecoin={pokecoin} setPokecoin={setPokecoin} />

      <Row className="justify-content-center text-center">
        <Col>
          <span css={style.mainTitle}>My Poke</span>
        </Col>
      </Row>

      {myListData && (
        <Row className="justify-content-center pt-4">
          {myListData.map((data, index) => (
            <Col key={index} xs={8} sm={6} md={4} lg={2}>
              <MyListCard
                data={data}
                myListData={myListData}
                setMyListData={setMyListData}
              />
            </Col>
          ))}
        </Row>
      )}

      {!myListData && (
        <Row className="justify-content-center text-center pt-4">
          <Col>
            <span css={style.text}>Empty! Roll your first Poke.</span>
          </Col>
        </Row>
      )}
    </>
  );
};

export default MyPokemonsList;
