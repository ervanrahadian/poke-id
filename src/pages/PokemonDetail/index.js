import { useQuery } from "@apollo/client";
import React from "react";
import { Button, Col, Image, Row } from "react-bootstrap";
import { useLocation } from "react-router-dom";
import { GET_POKEMON_DETAIL } from "../../graphql";
import {
  DetailBasics,
  DetailMoves,
  DetailStats,
  Loading,
  Toast,
} from "../../components";
import Swal from "sweetalert2";
import { checkDuplicateData, getFormattedText } from "../../utils";

/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";

const PokemonDetail = () => {
  const { state } = useLocation();
  const {
    loading: loadingDetailData,
    error: errorDetailData,
    data: detailData,
  } = useQuery(GET_POKEMON_DETAIL, {
    variables: {
      name: state,
    },
  });

  const style = {
    mainTitle: css({
      fontSize: "32px",
      fontWeight: "800",
    }),
    image: css({
      border: "4px solid #212529",
      borderRadius: "50%",
      height: "175px",
      width: "175px",
      transition: "0.2s",
      animation: "rotation 5s infinite linear",
      "@keyframes rotation": {
        "100%": {
          transform: "rotatey(360deg)",
        },
      },
    }),
    button: css({
      fontSize: "24px",
      height: "100px",
      width: "100px",
      borderRadius: "50%",
    }),
  };

  const swalWithStyle = Swal.mixin({
    customClass: {
      confirmButton: "btn btn-dark",
      cancelButton: "btn btn-outline-dark ms-3",
    },
    buttonsStyling: false,
  });

  const handleCatch = () => {
    const randomNum = Math.round(Math.random());

    if (randomNum === 0) {
      return swalWithStyle.fire({
        icon: "error",
        title: "Oops...",
        text: "Too hasty! Your Poke has run away.",
        confirmButtonText: "Got it",
      });
    } else {
      swalWithStyle
        .fire({
          icon: "success",
          title: "Congratulations...",
          text: "You caught the Poke!",
          input: "text",
          inputLabel: "Your Poke Nickname",
          confirmButtonText: "Add",
          showCancelButton: true,
          inputValidator: (value) => {
            if (!value) {
              return "Please add Nickname!";
            } else if (checkDuplicateData(value)) {
              return "Nickname already exist!";
            }
          },
        })
        .then((result) => {
          if (result.isConfirmed) {
            let myPokeData = [];

            if (localStorage.hasOwnProperty("myPoke")) {
              myPokeData = JSON.parse(localStorage.getItem("myPoke"));

              myPokeData.push({
                nickname: result.value,
                name: detailData.pokemon.name,
                image: detailData.pokemon.sprites.front_default,
              });

              localStorage.setItem("myPoke", JSON.stringify(myPokeData));

              return Toast({
                icon: "success",
                title: `${result.value} added!`,
              });
            } else {
              myPokeData.push({
                nickname: result.value,
                name: detailData.pokemon.name,
                image: detailData.pokemon.sprites.front_default,
              });

              localStorage.setItem("myPoke", JSON.stringify(myPokeData));

              return Toast({
                icon: "success",
                title: `${result.value} added!`,
              });
            }
          }
        });
    }
  };

  if (errorDetailData) {
    return Toast({
      icon: "error",
      title: "Something went wrong!",
    });
  }

  return (
    <>
      {loadingDetailData && <Loading />}

      {detailData && (
        <>
          <Row className="justify-content-center text-center">
            <Col>
              <span css={style.mainTitle}>
                {getFormattedText(detailData.pokemon.name)}
              </span>
            </Col>
          </Row>

          <Row className="justify-content-center text-center pt-4">
            <Col>
              <Image
                alt={getFormattedText(detailData.pokemon.name)}
                title={getFormattedText(detailData.pokemon.name)}
                src={detailData.pokemon.sprites.front_default}
                css={style.image}
              />
            </Col>
          </Row>

          <Row className="justify-content-center text-center pt-4">
            <Col>
              <Button variant="dark" css={style.button} onClick={handleCatch}>
                Catch
              </Button>
            </Col>
          </Row>

          <Row className="justify-content-center pt-4">
            <Col className="mb-4" md={4}>
              <DetailBasics data={detailData.pokemon} />
            </Col>

            <Col className="mb-4" md={4}>
              <DetailStats data={detailData.pokemon.stats} />
            </Col>

            <Col className="mb-4" md={4}>
              <DetailMoves data={detailData.pokemon.moves} />
            </Col>
          </Row>
        </>
      )}
    </>
  );
};

export default PokemonDetail;
