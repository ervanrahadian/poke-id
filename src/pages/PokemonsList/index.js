import { useLazyQuery, useQuery } from "@apollo/client";
import React, { useEffect, useState } from "react";
import { Button, Col, Row } from "react-bootstrap";
import { GET_POKEMONS_LIST, GET_TOTAL_POKEMONS_LIST } from "../../graphql";
import { ListCard, Loading, Microtransaction, Toast } from "../../components";

/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";

const PokemonsList = () => {
  const [pokecoin, setPokecoin] = useState(null);
  const {
    loading: loadingTotalData,
    error: errorTotalData,
    data: totalData,
  } = useQuery(GET_TOTAL_POKEMONS_LIST);
  const [
    getListData,
    { loading: loadingListData, error: errorListData, data: listData },
  ] = useLazyQuery(GET_POKEMONS_LIST);

  const style = {
    mainTitle: css({
      fontSize: "32px",
      fontWeight: "800",
    }),
    button: css({
      fontSize: "28px",
    }),
  };

  useEffect(() => {
    if (localStorage.hasOwnProperty("Pokecoin")) {
      setPokecoin(Number(localStorage.getItem("Pokecoin")));
    } else {
      const defaultPokecoin = Number(500);

      localStorage.setItem("Pokecoin", defaultPokecoin);
      setPokecoin(defaultPokecoin);

      return Toast({
        icon: "success",
        title: "500 Pokecoin added!",
      });
    }
  }, []);

  const handleRoll = () => {
    if (pokecoin < 100) {
      return Toast({
        icon: "error",
        title: "Pokecoin is not enough!",
      });
    } else {
      const max = totalData.pokemons.count - 5;
      const randomNum = Math.floor(Math.random() * (max + 1));
      const calc = Number(pokecoin - 100);

      getListData({
        variables: {
          limit: 5,
          offset: randomNum,
        },
      });

      localStorage.setItem("Pokecoin", calc);
      setPokecoin(calc);

      return Toast({
        icon: "success",
        title: "100 Pokecoin deducted!",
      });
    }
  };

  if (errorTotalData || errorListData) {
    return Toast({
      icon: "error",
      title: "Something went wrong!",
    });
  }

  return (
    <>
      {(loadingTotalData && <Loading />) || (loadingListData && <Loading />)}

      <Microtransaction pokecoin={pokecoin} setPokecoin={setPokecoin} />

      <Row className="justify-content-center text-center">
        <Col>
          <span css={style.mainTitle}>Roll And Catch</span>
        </Col>
      </Row>

      {!loadingListData && !listData && (
        <Row className="justify-content-center text-center pt-4">
          <Col>
            <Button variant="dark" css={style.button} onClick={handleRoll}>
              Roll (100 Pokecoin)
            </Button>
          </Col>
        </Row>
      )}

      {listData && (
        <>
          <Row className="justify-content-center pt-4">
            {listData.pokemons.results.map((data) => (
              <Col key={data.id} xs={8} sm={6} md={4} lg={2}>
                <ListCard data={data} />
              </Col>
            ))}
          </Row>

          <Row className="justify-content-center text-center pb-4">
            <Col>
              <Button variant="dark" css={style.button} onClick={handleRoll}>
                Reroll (100 Pokecoin)
              </Button>
            </Col>
          </Row>
        </>
      )}
    </>
  );
};

export default PokemonsList;
