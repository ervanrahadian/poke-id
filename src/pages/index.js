import MyPokemonsList from "./MyPokemonsList";
import PokemonDetail from "./PokemonDetail";
import PokemonsList from "./PokemonsList";

export { MyPokemonsList, PokemonDetail, PokemonsList };
