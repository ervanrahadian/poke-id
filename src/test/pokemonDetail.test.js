import {
  ApolloClient,
  ApolloProvider,
  HttpLink,
  InMemoryCache,
} from "@apollo/client";
import { render } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { PokemonDetail } from "../pages";

const client = new ApolloClient({
  link: new HttpLink({
    uri: "https://graphql-pokeapi.vercel.app/api/graphql",
  }),
  cache: new InMemoryCache(),
});

test("check detail page", () => {
  render(
    <ApolloProvider client={client}>
      <MemoryRouter initialEntries={[{ pathname: "/pokemon-detail" }]}>
        <Routes>
          <Route path="/pokemon-detail" element={<PokemonDetail />} />
        </Routes>
      </MemoryRouter>
    </ApolloProvider>
  );

  expect(render.toString()).toMatchSnapshot();
});
