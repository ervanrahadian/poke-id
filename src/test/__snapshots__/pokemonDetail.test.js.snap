// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check detail page 1`] = `
"function render(ui, {
  container,
  baseElement = container,
  queries,
  hydrate = false,
  wrapper: WrapperComponent
} = {}) {
  if (!baseElement) {
    // default to document.body instead of documentElement to avoid output of potentially-large
    // head elements (such as JSS style blocks) in debug output
    baseElement = document.body;
  }

  if (!container) {
    container = baseElement.appendChild(document.createElement('div'));
  } // we'll add it to the mounted containers regardless of whether it's actually
  // added to document.body so the cleanup method works regardless of whether
  // they're passing us a custom container or not.


  mountedContainers.add(container);

  const wrapUiIfNeeded = innerElement => WrapperComponent ? /*#__PURE__*/React.createElement(WrapperComponent, null, innerElement) : innerElement;

  (0, _actCompat.default)(() => {
    if (hydrate) {
      _reactDom.default.hydrate(wrapUiIfNeeded(ui), container);
    } else {
      _reactDom.default.render(wrapUiIfNeeded(ui), container);
    }
  });
  return {
    container,
    baseElement,
    debug: (el = baseElement, maxLength, options) => Array.isArray(el) ? // eslint-disable-next-line no-console
    el.forEach(e => console.log((0, _dom.prettyDOM)(e, maxLength, options))) : // eslint-disable-next-line no-console,
    console.log((0, _dom.prettyDOM)(el, maxLength, options)),
    unmount: () => {
      (0, _actCompat.default)(() => {
        _reactDom.default.unmountComponentAtNode(container);
      });
    },
    rerender: rerenderUi => {
      render(wrapUiIfNeeded(rerenderUi), {
        container,
        baseElement
      }); // Intentionally do not return anything to avoid unnecessarily complicating the API.
      // folks can use all the same utilities we return in the first place that are bound to the container
    },
    asFragment: () => {
      /* istanbul ignore else (old jsdom limitation) */
      if (typeof document.createRange === 'function') {
        return document.createRange().createContextualFragment(container.innerHTML);
      } else {
        const template = document.createElement('template');
        template.innerHTML = container.innerHTML;
        return template.content;
      }
    },
    ...(0, _dom.getQueriesForElement)(baseElement, queries)
  };
}"
`;
